### DISCUSSION TITLE
Universal listings (search and filtering)

### DISCUSSION THREAD

**Discussion by benenright** [wagtail/wagtail#10446] [Ideas]
### Summary
Exploring design solutions for making sites / sections with 1000s of pages and little to no tree structure easier to navigate in the CMS.

The aim is to enhance the searching / filtering capabilities of the main page listing view (and to eventually roll out the same patterns and functionality to all listings within the CMS).

### Background
- See https://github.com/wagtail/rfcs/pull/82
- See https://github.com/wagtail/roadmap/issues/33

### Considerations
- An accessible and scalable mechanism for filtering
- How to show when filters/search terms are applied
- Approach to fetching results and impact on performance
- Adding additional information such as 'X results found' based on filtering/search criteria
- Can users 'save' a filtered view to quickly find it again in the interface
- UI patterns that can be consistent across all listings (not just Pages)
- Impact on the listing item information (eg. show no. of direct descendants, show location in the tree in search results)
- Impact on the listing item styling/actions (eg. reduce vertical height, improve a11y of the actions (Edit, Add child page etc), add a folder icon to help identify parent pages)
- Impact on the role and functionality of the fly-out page explorer

####  Default page listing
<img width="1512" alt="listing-pages" src="https://github.com/wagtail/wagtail/assets/4907819/f7e1e585-1054-4072-a977-2a0dc705cc51">

####  Search results
<img width="1512" alt="search-results" src="https://github.com/wagtail/wagtail/assets/4907819/12a685d9-b1e5-46ba-92f5-7d585463bd1a">

####  Item hover state
<img width="1325" alt="Screenshot 2023-05-16 at 13 04 05" src="https://github.com/wagtail/wagtail/assets/4907819/e57f4a7b-8576-45f8-bfe4-8c53cd5b298b">

####  Filter dialogs
<img width="1512" alt="default-filter-dialog" src="https://github.com/wagtail/wagtail/assets/4907819/efd0dcb0-fce7-48e5-a10d-04b3126a7cc3">

<img width="475" alt="filter-category" src="https://github.com/wagtail/wagtail/assets/4907819/03156970-5b6c-4954-a7d6-dda5e61170df">

<img width="443" alt="filter-date-range" src="https://github.com/wagtail/wagtail/assets/4907819/eae5aee7-bc7f-420b-b7ff-0a721c57d5d3">

<img width="438" alt="filter-owner" src="https://github.com/wagtail/wagtail/assets/4907819/45e6ccf8-efb9-4bd0-8faa-5c13c3d7138f">

####  Filter pills
<img width="1082" alt="filter-pill" src="https://github.com/wagtail/wagtail/assets/4907819/a967d642-f340-44de-a4b7-8b2dad7cd5c1">

<img width="1132" alt="selected-filter-category" src="https://github.com/wagtail/wagtail/assets/4907819/59b4776a-be82-4244-b1b6-0ff880198809">

<img width="1312" alt="multiple-filters" src="https://github.com/wagtail/wagtail/assets/4907819/ffbb17ab-a1a3-4b3f-a3a6-68148bf72f6f">

####  Save view
<img width="1512" alt="save-view" src="https://github.com/wagtail/wagtail/assets/4907819/631c4fb2-9e2d-44a0-9920-c98e6bdc8a7f">

<img width="1512" alt="saved-views-select" src="https://github.com/wagtail/wagtail/assets/4907819/38e9694a-4841-4282-8bca-de1b311f10a6">

### Open questions

- What is the best approach for the visibility of the listing item actions? It feels heavy in the UI to have an omni-present 'Edit' button, and since this action can also be achieved by clicking on the item title, there is an argument for the 'Edit' button to appear on hover
- Whether we introduce the 'Add child page' shortcut action to each item, where child pages are allowed to be added
- Impact on the role and functionality of the fly-out page explorer - and whether we consider this as part of this work
- Impact on Chooser modals

Welcome all feedback and thoughts people would like to contribute.

**Comment by pySilver** [comment]
That looks really really good!

**Comment by tbrlpld** [comment]
Yes please!

**Comment by enzedonline** [comment]
Superb :)

Request for multi-lingual sites - please add 'results limited to current listed locale' in considerations. Thanks!

**Reply by benenright** [reply]
Good idea, will see where this feels natural. It does open the question around whether we want to display the current locale upfront on the listing view (it's currently now hidden away in the status panel). But if it were more visible, it may negate the need to display  'results limited to current listed locale' since that may be more apparent.

**Comment by ababic** [comment]
This is looking superb @BenEnright! Especially loving the filters 😍! 

A couple of thoughts that occurred to me are:

1. What is the default action when clicking on the page title? Will that continue to always go to the explorer view for that page? 
2. Do we have any research on the actions most performed from the explorer listing? I feel like the "edit" action should certainly be easier to see and access. Would icon-only buttons work for the key actions? 
3. Have you played with repurposing the right-most column for actions? I feel like a series of small buttons might work nicely there (instead of next to the page title) as it doesn't disrupt the visual flow of the row as much. The "explore this section" action feels a little less important outside of the tree view, but is using up prime real estate.

**Reply by benenright** [reply]
Hi Andy

1. Yea, the default action when clicking a page would stay as 'Edit'
2. Agree 'Edit' should be easy to access, but since the default action when clicking a page title is 'Edit', we were considering still keeping the _additional_ Edit button only appear on hover (to reduce clutter). Icon-only buttons have been considered and maybe we could revisit... although there was something quite strange about the edit 'pen' icon on it's own without a label :). 
3. Yes we considered moving the actions further to the right of the table (I think we can try this again based on your thoughts about row disruption). A reason we kept them nearer the page title was that it felt like in general it was an easier place on the screen to get to (less distance to move for mouse users) when scrolling down the table (where in general we expect people to be hovering more on the left side while reading the page titles). We still need a place for the 'explore this section' button, and the paradigm of the right arrow (on the far right side) to dive down a level still feels natural (akin to mobile menus etc.)

**Comment by cnk** [comment]
I like this! I have a site that doesn't have a particularly main content tree but we do have years of imported news and events pages so we have built custom versions of what you are proposing here (though not as nice because I don't have your design skilz). From that perspective, it would be great if the filtering component were configurable since sites like ours will have immediate uses for custom filters - and because all snippet listings that eventually use this will want to set up their own filtering options.

**Comment by ababic** [comment]
Just thought of another challenge that isn't explicitly addressed in the designs so far...

What does the switch between "only show children of the current explorable page" and "show matches from anywhere" look like? And how does that work for multisite / translated projects?

Maybe we need two completely separate "Page explorer" and "All pages" menu items, and the latter would go to somewhere with a completely different header and site / language filters?

**Reply by benenright** [reply]
Yea, I think we could still consider how the search/filter can be either "show any descendants below this point in tree" or "show matches from anywhere". Where in the latter we will need to find a way to show locale for each result (and also consider the need for a 'locale/language' filter).

However, so far our mvp version of this is not to create a seperate view for 'All pages' but rather to enhance the existing tree view that people are familiar with. In our designs we are currently limited to showing results 'below' your point in tree - which in most cases I believe negates the need to show different locales because (unless you are at Root) in most cases you are already within a specific locale.

@gasman @thibaudcolas may want to chime in on this?

**Reply by thibaudcolas** [reply]
Yup. In a tree structure world, the "treeless" experience of "show matches from anywhere" will essentially amount to applying filters from the root level. Which is accessible from the top of the page explorer. It’d be nice if this was more easily accessible perhaps.

**Reply by ababic** [reply]
> the search/filter can be either "show any descendants below this point in tree" or "show matches from anywhere".

Okay. In the explorer view, we're currently showing only 'direct children' of the current point in the tree. So, I think automatically playing with that rule when searching or filtering could cause a bit of confusion. I would suspect we need to stick with showing direct children in both cases by default, but have some opt-in mechanism for showing more 'levels' (e.g.  a toggle switch or checkbox).

> the "treeless" experience of "show matches from anywhere" will essentially amount to applying filters from the root level. Which is accessible from the top of the page explorer. It’d be nice if this was more easily accessible perhaps.

Okay, one thing to bare in mind there is: There's not really any reason for most editors to care about the root page at all, as content is never at that level. So, I worry it wouldn't be an obvious place for them to navigate to, even if that option were more visible. A design challenge, for sure.

If similar options were available on the general/global 'Search' view too, the "View results from anywhere" option could just link out to that, potentially?

**Comment by thibaudcolas** [comment]
Just wanted to capture another use case we discussed with colleagues today (cc @helenb) – for QA purposes, it’s *really* common to need to find a page of a given page type. And potentially doing that for multiple page types.

I believe the current designs as proposed would cater for that scenario well – go to the root level, filter by page type, pick the first one. But it’s such a common scenario I thought it’d be worth mentioning anyway.

**Reply by ababic** [reply]
I added this fly-out menu to Tate using ModelAdmin, which they love. It would be awesome if Wagtail could roll it's own version of this:

<img width="1506" alt="Screenshot 2023-05-30 at 9 50 58 pm" src="https://github.com/wagtail/wagtail/assets/7779588/768828b7-c5ca-435c-8dba-6dadf38b4e49">

**Reply by ababic** [reply]
Just spitballing here: If we DID happen to have a separate "View all" option for pages, the "View by type" options could simply link to a pre-filtered version of that.

**Reply by easherma-truth** [reply]
Its also common (IMO) to be able to search by page slug and not just title or page type 🤔

**Reply by pySilver** [reply]
And filter by Site

**Comment by SebCorbin** [comment]
Another use case from us, we have a site where the page tree is not very relevant: some editorial pages like "Legal mentions" and another page-based model (a bit like landing pages) for which we want page features like workflow, previewing, and so on, along with our business configuration.

In our case the Page explorer is too far-fetched for this kind of site structure so we would love to get rid of it entirely. So far we are stuck with it, but by using ModelAdmin we have a kind of basic list so this feature would help a lot. 
A thing to note, the menu breadcrumb (which item is active) would need a bit of thinking, do we want to retain the path (list) from which we come when editing a Page ?

**Comment by jhonatan-lopes** [comment]
Hi folks, I've just submitted a [PR](https://github.com/wagtail/wagtail/pull/10850) for a new built-in report focusing on page types, which answers the following questions:

- Which page types do we have on our CMS?
- How many pages of that page type do we have?
- When was a page of that type last edited? By whom? Which page was that?

It can also show a list of all the pages of a particular type.

**Comment by benenright** [comment]
Hi

During development of the new listings, some UX issue have come to light so we have adjusted some of the UI.

**Notes**
- Having the search and filtering bar underneath creates a very tall header and since this header is sticky (to enable consistent access to and behaviour of the side panels), we have created a version where we move the search and filter button up alongside the panel icons, to achieve a shorter sticky header (details below)
- We also reduced the size of the page title (which is also the 'current page' in the collapsed breadcrumb)
- The filter 'pills' will still appear in a row underneath, making the header taller

<img width="1512" alt="Screenshot 2023-10-17 at 12 08 11" src="https://github.com/wagtail/wagtail/assets/4907819/745d932e-fa1b-4e47-826d-203940faaf7a">

<img width="1512" alt="Screenshot 2023-10-17 at 12 11 41" src="https://github.com/wagtail/wagtail/assets/4907819/a7256ae0-4c39-45a0-afaf-570c7d974b41">

<img width="757" alt="Screenshot 2023-10-17 at 12 12 24" src="https://github.com/wagtail/wagtail/assets/4907819/35f258ce-b171-43a7-9e55-a2c564cfaca8">

<img width="1512" alt="Screenshot 2023-10-17 at 12 13 06" src="https://github.com/wagtail/wagtail/assets/4907819/9496de4e-7e8c-4322-b2a7-53eb3e38198a">

**General listing header would share similar patterns:**
![general listing](https://github.com/wagtail/wagtail/assets/4907819/cbc10983-a7c0-46d5-a116-6c63dfc19528)

**Comment by gasman** [comment]
Just a note that a request has come up to add "is a parent page or not" to the set of filtering options. (That should be a simple one to implement - filter on the 'numchild' field being greater than 0.)

**Comment by thibaudcolas** [comment]
Discussed this week with Ben – it’d be nice to have a way to view the "treeless" version of a specific view without having to trigger a search or applying a filter. For example, if I’d like to see all of the site’s pages.

**Reply by Scotchester** [reply]
Assuming it's all controlled by query string params in the URL, perhaps a hook could be used to add permanent menu links pointing to specific views? These could have specific search queries, or filters applied, or neither.

For bonus points, a way for users to add _their own_ saved views to the sidebar for easy access would be pretty cool.

**Reply by thibaudcolas** [reply]
Ah yep! Since the sidebar is for navigation, it’d probably be a very natural place for those "saved views" / bookmarks.

**Comment by enzedonline** [comment]
Not sure if this is the right place for this - 

It would be a great boost in UX if the current view settings were retained after performing an action on the listing page. This applies to all the listing views I've seen so far (i.e. pages, snippets, users, forms ...).

Currently, after performing an action, the returned page is the first page of the default view. This can be quite frustrating when you've got a filtered, ordered or paginated view you're working with.

The view could be all pages created after X, user list sorted by last login date or page x of submitted forms for instance. Ideally, you'd want all those attributes of the view retained after performing the menu action (delete, unpublish etc.).

**Reply by thibaudcolas** [reply]
Interesting, thank you. Do you have examples of other tools that work like this? This would make perfect sense to me if those operations were AJAX-based, but when moving from page to page I can imagine it would be confusing if you returned to a filtered view.

**Reply by cnk** [reply]
Actually the PR my team submitted to add `next_url` to most page actions was so we could do exactly what @enzedonline suggested. We have a snippet index view where our editors search for items to review and edit. We had multiple requests for "can't this go back to my search results?" so we customized that listing to return to the filtered list.

**Reply by enzedonline** [reply]
I think people are far more used to AJAX type behaviour these days, there's not much around that still does a refresh for list actions like this. Feedback is definitely that the expected behaviour is to keep the view you're already on, not have it reset each time.

The best example I can think of are the more old-school webmail interfaces (c-panel type thing) - sort email by name/date, be on page x of items. If you were to delete an email on that for example, you wouldn't want your inbox view to return to the first page sorted by newest first, you'd expect you view to stay as you were.

**Reply by laymonage** [reply]
Thank you, it looks like we have this behaviour in some places but not all. For example, the documents index view does this:

https://github.com/wagtail/wagtail/blob/b7beb36df6bc89e6152eae6c7d2dabaee7bd660a/wagtail/documents/templates/wagtaildocs/documents/list.html#L42-L44

https://github.com/wagtail/wagtail/blob/b7beb36df6bc89e6152eae6c7d2dabaee7bd660a/wagtail/documents/views/documents.py#L78-L81

If we want to keep this behaviour, we'll need to make sure it's applied everywhere.

**Reply by thibaudcolas** [reply]
What do you all think of #9177?

**Comment by jsma** [comment]
Hi all, 

Per the [Migrating from ModelAdmin to Snippets](https://docs.wagtail.org/en/v5.2.3/reference/contrib/modeladmin/migrating_to_snippets.html#using-modeladmin-to-manage-page-models) guide:

> ModelAdmin allows the registration of Page models, but the main use case is to create custom page listing views. The create and edit views are not supported by ModelAdmin, as there are page-specific operations in those views that are best handled by Wagtail’s page views. For this reason, registering a Page model as a snippet is not supported.
> In a future release, Wagtail will introduce a new “treeless” listing view for pages, as outlined in [RFC 082: Treeless page listings](https://github.com/wagtail/rfcs/pull/82) and the [Universal Listings discussion](https://github.com/wagtail/wagtail/discussions/10446). This feature will allow for custom page listing views and will be the recommended way to achieve this use case. Feedback to this upcoming feature is welcome.

I'm not clear where we're at based on this thread. I can see that it's possible to filter by page type in Universal Listings but it's not clear to me how one would replicate the ModelAdmin functionality of creating a custom menu item e.g., Marketing -> Blog Posts so that staff can jump right to the page types that matter to them. Another use case we could handle with ModelAdmin was to add custom filters to this index view. For example, the Blog Posts ModelAdmin might have Blog Category filters on the right. With Universal Listings, the filters appear to be limited to fields that are common to all page types.

Is the solution to make a custom view/template and wire up the URL(s) using `register_admin_menu_item`? Previously it was a lot more clear to me how to add a mix of items to a custom menu. Now I'm thoroughly confused how we're supposed to wire things together if we have a custom menu item that may include models registered with SnippetViewSet and some with ModelViewSet plus custom views with `register_admin_menu_item`. Do we put them all in a ModelViewSetGroup or SnippetViewSetGroup or should we not mix items?

**Reply by laymonage** [reply]
> I'm not clear where we're at based on this thread. I can see that it's possible to filter by page type in Universal Listings but it's not clear to me how one would replicate the ModelAdmin functionality of creating a custom menu item e.g., Marketing -> Blog Posts so that staff can jump right to the page types that matter to them. Another use case we could handle with ModelAdmin was to add custom filters to this index view. For example, the Blog Posts ModelAdmin might have Blog Category filters on the right. With Universal Listings, the filters appear to be limited to fields that are common to all page types.

This was added in #11485, which will be in 6.1. A follow-up to allow creating a page from such a list is #11734 (if we can get a PR in before feature freeze, that'll be in 6.1 as well).

> Is the solution to make a custom view/template and wire up the URL(s) using `register_admin_menu_item`? Previously it was a lot more clear to me how to add a mix of items to a custom menu. Now I'm thoroughly confused how we're supposed to wire things together if we have a custom menu item that may include models registered with SnippetViewSet and some with ModelViewSet plus custom views with `register_admin_menu_item`. Do we put them all in a ModelViewSetGroup or SnippetViewSetGroup or should we not mix items?

I wouldn't recommend mixing the items, but I believe there isn't really anything that prevents you from doing so. The safest option would be to use the `ViewSetGroup` base class:

https://github.com/wagtail/wagtail/blob/77d783b397d023d2d47ddc8157634e96d3995a1d/wagtail/admin/viewsets/base.py#L134-L143

All the `ModelViewSetGroup` does is add a reasonable default menu label if one isn't supplied:

https://github.com/wagtail/wagtail/blob/77d783b397d023d2d47ddc8157634e96d3995a1d/wagtail/admin/viewsets/model.py#L662-L679

And the `SnippetViewSetGroup` only sets `menu_item_is_registered` to the items to ensure the Snippets menu item is properly hidden when all items have their own menu items:

https://github.com/wagtail/wagtail/blob/77d783b397d023d2d47ddc8157634e96d3995a1d/wagtail/snippets/views/snippets.py#L1290-L1310

So, as long as you have the `menu_label` set for the menu group, and you set `menu_item_is_registered = True` in your `SnippetViewSet`s (or leave it if you don't care about hiding the Snippets menu item), you should be OK with using the base `ViewSetGroup`.

Or, alternatively, you might be able to use `SnippetViewSetGroup`. I think the code should handle non-snippet/non-model items gracefully (and if not, we should probably change that).

**Comment by thibaudcolas** [comment]
Tracking the progress to date on universal listings:

- https://github.com/wagtail/roadmap/issues/33
- https://github.com/wagtail/roadmap/issues/62
- https://github.com/wagtail/roadmap/issues/75

Here is the work we believe still remains. Sharing this as a table rather than GitHub issues as we’ve not specified exactly what needs to happen beyond the high-level task label

| Status | Task                                                                | Assignee   | Estimate (hours) |
| ------ | ------------------------------------------------------------------- | ---------- | ---------------- |
| ✅    | Locked pages report                                                 | @laymonage | 4                |
| ✅    | Workflows report                                                    | @laymonage | 4                |
| ✅    | Workflow tasks report                                               | @laymonage | 4                |
| ✅    | Site history report                                                 | @laymonage | 4                |
| ✅    | Aging pages report                                                  | @laymonage | 4                |
| ✅    | Page types usage report                                             | @laymonage | 4                |
| ✅   | Workflows usage view                                                | @laymonage | 4                |
| ✅    | Workflow history view                                               | @laymonage | 6                |
| ✅   | Workflow history detail view                                        |@laymonage            | 3                |
| Todo   | Mobile support                                                      |            |                  |
| Todo   | Saved views                                                         |            | 20               |
| Todo   | Pages: secondary actions in kebab menu                              |            | 4                |
| Todo   | Pages: 'navigate to children' control should show descendant counts |            | 16               |
| Todo   | Improvements to making filters pluggable (active filters pills)                              |            | 16               |
| ✅   | Images add/edit view                                                | @laymonage           | 4                |
| ✅   | Sites add/edit view                                                 |  @laymonage          | 2                |
| ✅   | Locales add/edit view                                               | @laymonage           | 2                |
| ✅   | Collections add/edit view                                           |  @laymonage           | 3                |
| ⌛️   | Redirects add/edit view                                             |            | 8                |
| v6.3\*   | Search promotions add/edit view                                     |            | 6                |
| v6.3\*   | Settings view                                                       |            | 4                |
| Todo   | Add slim header to styleguide                                       |            | 2                |


@laymonage is keen to do quite a bit of this but we’re leaving open the exact time frame for now.

**Comment by thibaudcolas** [comment]
👋 for anyone following this thread – we’d be interested in your thoughts on the __Saved views__ feature hinted at in existing design prototypes: 

<img width="1512" alt="save-view" src="https://github.com/wagtail/wagtail/assets/4907819/631c4fb2-9e2d-44a0-9920-c98e6bdc8a7f">

<img width="1512" alt="saved-views-select" src="https://github.com/wagtail/wagtail/assets/4907819/38e9694a-4841-4282-8bca-de1b311f10a6">

---

We’ve done the UX prototyping visible in those mockups, and some thinking about how to proceed with implementation, but decided to leave it there to focus on more fundamental aspects of universal listings. For the upcoming Wagtail 6.3\* release, Sage is working on additional "[Universal Design](https://github.com/wagtail/roadmap/issues/88)" technical and UX debt cleanup. Afterwards, we have no specific plans.

Do you think "Saved views" as enough potential to formalise as a feature request? If not it’ll remain as a design concept for future consideration.

**Reply by SebCorbin** [reply]
That's a feature that would make Wagtail an admin on steroids, we would certainly love it.

From the mockup I see three things:
 - it would gracefully integrate with the Page explorer, although that's something we've entirely removed in our case to make use of the new "Universal listing" (with a "flat" structure). That means the "Location" part would not always be relevant.
 - another question (but more on the implementation side) comes to me: how would an user delete an existing view once selected?
 - and finally, have you worked on a mockup integrated with the `slim_header`?

**Reply by zerolab** [reply]
A few other things that come to mind:
- can the saved views be edited after creation?
- are they global or per user? if global, it may be worth considering the implications of "user A saves a view", then "user B changes or removes the view"
- are actions (save view, etc) going to the audit log?

**Reply by Stormheg** [reply]
This looks very useful! I echo the points Dan mentioned already: how to edit/delete a saved view? and are they global?

Also, can we as developers create saved views (or some equivalent)? Imagine an `EventPage` listing where people would commonly want to filter for events in the future. It would be helpful if there was a quick way to set filters that way.

Edit: can we also save a certain ordering? To always show pages sorted by recent changes for example.

**Reply by thibaudcolas** [reply]
I think we asked ourselves the same thing :) editing / deleting seemed straightforward to add. For "global or per user", we discussed both, don’t remember there being a clear winner.

**Reply by ababic** [reply]
From an implementation perspective, it should be relatively easy to silently store an `owner` for each item (which Wagtail does for a few other things), and provide a checkbox in the edit form to make certain items available to everyone. As an MVP, I think it would be fine to only allow the `owner` to edit their own items.

**Comment by tbrlpld** [comment]
The mock-up contains a selector for the page size, it seems like that did not make it into the implementation yet. It that still planned? 

Also, Wagtail's standard pagination component is pretty bare, only "next" and "previous". Are there any plans to make that more feature-full (e.g. elided, direct page entry, ...)?

**Reply by zerolab** [reply]
@tbrlpld See https://github.com/wagtail/wagtail/issues/12004

**Reply by tbrlpld** [reply]
Thanks @zerolab.
